cfg_mgmt/btp_application_certificate.py

```diff
@@ -0,0 +1,256 @@
+        oid = _str_to_names[name]
+        attributes.append(x509.NameAttribute(oid, value))
+    csr = x509.CertificateSigningRequestBuilder().subject_name(
+        x509.Name(attributes)
+    ).sign(key, hashes.SHA256())
+    csr_pem = csr.public_bytes(serialization.Encoding.PEM).decode('utf-8')
+    return csr_pem, key_pem
+
+
+def _extract_client_certificate(cert_response: dict) -> str:
+    pkcs7_pem = cert_response['value']
+    certs = pkcs7.load_pem_pkcs7_certificates(pkcs7_pem.encode('utf-8'))
+    if len(certs) == 0:
+        raise ValueError('no certificates found in response')
+    return certs[0].public_bytes(serialization.Encoding.PEM).decode('utf-8')
+
+
+def rotate_cfg_element(
+    cfg_element: BtpApplicationCertificate,
+    cfg_factory: model.ConfigFactory,
+) -> typing.Tuple[cfg_mgmt.revert_function, dict, model.NamedModelElement]:
+    gbaas_auth = cfg_factory.btp_application_certificate(cfg_element.auth_application_certificate())
+    gbaas_client = GBaasAppClient(gbaas_auth)
+
+    # calc next serial no
+    cn = cfg_element.common_name()
+    serial_no, base = BtpApplicationCertificate.get_serial_no_from_common_name(cn)
+    if serial_no == 0:
+        raise ValueError(f'unexpected cn: {cn}')
+    next_sn = serial_no + 1
+    for info in gbaas_client.list_certificates_by_base(base):
+        if info.serial_no >= next_sn:
+            next_sn = info.serial_no + 1
+    next_cn = f'{next_sn}.{base}'
+
+    # create certificate
+    csr_pem, key_pem = _create_csr(cfg_element.subject(next_cn))
+    sb_auth = cfg_factory.btp_service_binding(cfg_element.cert_service_binding())
+    cs_client = CertServiceClient(sb_auth.credentials())
+    response = cs_client.get_client_certificate_chain(csr_pem, cfg_element.validity_in_days())
+    cert_pem = _extract_client_certificate(response)
+
+    # add certificate to GBaas application
+    id = gbaas_client.put_certificate(
+        cert_pem=cert_pem,
+        desc=f'CN={next_cn}',
+        scopes=cfg_element.scopes(),
+    )
+
+    secret_id = {'common_name': cn}
+    raw_cfg = copy.deepcopy(cfg_element.raw)
+    raw_cfg['certificate_pem'] = cert_pem
+    raw_cfg['private_key_pem'] = key_pem
+    raw_cfg['common_name'] = next_cn
+    updated_elem = BtpApplicationCertificate(
+        name=cfg_element.name(), raw_dict=raw_cfg, type_name=cfg_element._type_name
+    )
+
+    def revert():
+        gbaas_client.delete_certificate(next_cn, id)
+
+    return revert, secret_id, updated_elem
+
+
+def delete_config_secret(
+    cfg_element: BtpApplicationCertificate,
+    cfg_queue_entry: cmm.CfgQueueEntry,
+    cfg_factory: model.ConfigFactory,
+):
+    logger.info('Deleting old certificates')
+    gbaas_auth = cfg_factory.btp_application_certificate(cfg_element.auth_application_certificate())
+    gbaas_client = GBaasAppClient(gbaas_auth)
+    cn = cfg_queue_entry.secretId['common_name']
+    serial_no, base = BtpApplicationCertificate.get_serial_no_from_common_name(cn)
+    if serial_no == 0:
```

Comment:
if not serial_no?

Reply:
also: would it make sense to already validate/raise in `get_serial_no....` method?

Reply:
done